Assign Copilot during `gh issue create`

andyfeller opened the pull request:
Fixes github/cli#920

This pull request expands `gh issue create` to support assigning Copilot during issue creation in multiple ways:

1. Interactively through `Add metadata` experience
2. Non-interactively through `gh issue create --assignee @copilot`
3. Web browser through `gh issue create --assignee @copilot --web`

## Underlying changes that affect how `gh pr create` works but not the user experience

`gh issue create` and `gh pr create` used a custom bit of logic for looking up information to resolve metadata:

https://github.com/cli/cli/blob/64191577e4ae5974b1741960fa481f192cc2f3c4/api/queries_repo.go#L1045-L1142

This logic needed to change to support Copilot and other bots.  Additionally, there was a concern about inconsistent approaches to retrieving this data and the affect it would have on the 2 commands.

As such, this pull request includes a minor refactor to retrieving data to resolve metadata without changing the `gh pr create` experience.

## Work remaining

- [x] Review and expand existing `gh issue create` tests around `assignableUsers` on GHES versus `suggestedActors` on GHEC
- [x] Review and expand existing `gh pr create` tests around changes around resolving metadata
- [ ] Post demonstration of new capability

andyfeller committed:
Initial assign Copilot during issue create changes

- updated command documentation for new `--assignee @copilot` syntax
- updated `gh issue create` logic with feature detection for actor availability
- fixed bug where `--assignee` would not select default assignees due to login vs display name mismatch
- updated survey prompt logic to avoid parsing prompt options in favor of correlating with assignee / actor source

One thing that is not included in these changes is incorporating the GraphQL mutation to replace assignees with actors, which I will continue iterating on.

andyfeller committed:
Add non-TTY issue create copilot test, todos

andyfeller committed:
Resolve issue and PR metadata consistently

This commit refactors how `gh issue create` and `gh pr create` retrieve information needed to resolve metadata to be more in line with the approach used in `gh issue edit` and `gh pr edit`.

Previously, both commands used `prshared.fillMetadata(...)` function to retrieve assignees, reviewers, labels, and teams outside of `api.RepoMetadata(..)`.  Now, these commands will consistently use the same logic and data for resolving metadata.

andyfeller committed:
Fix `gh issue create` tests, copilot web assign

This commit is primarily focused on fixing the existing `gh issue create` tests with the changes to the underlying queries retrieving the data for resolving metadata.

Additionally, a new test case for `gh issue create --web --assignee @copilot` was added to mirror existing tests.  While exercising this capability, I found out the web UI apparently wants `copilot` instead of the `copilot-swe-agent` login, so changes to address that were a part of this commit.

andyfeller committed:
Fix `gh pr create` tests from metadata change

These are the minimum fixes necessary to repair `gh pr create` tests after refactoring the underlying metadata resolution logic.

andyfeller commented on pkg/cmd/issue/create/create.go:
```diff
@@ -166,20 +176,29 @@ func createRun(opts *CreateOptions) (err error) {
 		return err
 	}
 
+	if issueFeatures.ActorIsAssignable {
+		assignees = copilotReplacer.ReplaceSlice(assignees)
+	}
```
**thought:** not sure we need to conditionalize this based on actor support 🤔

andyfeller replied:
@BagToad : I'd appreciate your $0.02 and recommendation here. 🙇

BagToad replied:
Hmmm

If we don't conditionalize on actor support:

- then it continues to work as-is for github.com, and
- it continues to _fail_ when a user attempts to use `@copilot` on a system where it's not supported, **but** since we are replacing the name even then, they'll get a different GraphQL error message back that references the replaced name instead of what they entered. 

Because of that different error message, I'd lean towards not replacing the name unless there is actor support. 

I want the user to get an error message along the lines of "invalid user `@copilot`" instead of "invalid user `copilot-swe-agent`.

Making some assumptions about how error messages bubble up here, but let me know if this is enough to make a decision 👍

andyfeller replied:
Well reasoned, thanks! 🙇 Let's keep this as-is then.

BagToad commented on pkg/cmd/pr/shared/params.go:
```diff
@@ -314,17 +286,25 @@ func (r *MeReplacer) ReplaceSlice(handles []string) ([]string, error) {
 // CopilotReplacer resolves usages of `@copilot` to Copilot's login.
-type CopilotReplacer struct{}
+type CopilotReplacer struct {
+	webMode bool
```
thought: I was mulling over how this is called `webMode` which does a good job of indicating when you'd want to use this property, but less so in communicating what it does. It made me wonder if something like `useDisplayName` might be more descriptive of the behavior. That comes at a trade-off of not really communicating when you'd want to use that, but perhaps it could be supplemented by a code comment. 

I'll leave it up to you to decide if you like that idea or not 🙂 Either way, feel free to just resolve this comment when you've read it.

andyfeller replied:
That's a very good point and I'm open to rename it. 🤔  Since Copilot's display name is `Copilot (AI)`, would there be confusion or errors by using `displayName` language?

BagToad commented on pkg/cmd/pr/shared/params.go:
```diff
@@ -314,17 +286,25 @@ func (r *MeReplacer) ReplaceSlice(handles []string) ([]string, error) {
+		return "copilot"
 	}
-	return handle
+	return api.CopilotActorLogin
```
question: do you think there would be value in returning an exported constant like `api.CopilotActorDisplayName`? 

I'm not convinced there's value in it, which is why this is a question and not a suggestion. My intuition is to desire consistency with the existing return value, but that's pretty much the only motivation behind me bringing this up 🤪

andyfeller replied:
I'm open to the idea.  How would this work given it's display name is `Copilot (AI)` in the existing experience?

BagToad commented on pkg/cmd/pr/shared/survey.go:
```diff
@@ -199,9 +202,27 @@ func MetadataSurvey(p Prompt, io *iostreams.IOStreams, baseRepo ghrepo.Interface
 	}
+
+	// In order to select any provided assignee, the following code needs to take the provided default assignees in `state`
+	// and translate them to the appropriate actor / user.
```
Suggestion (if-minor):

I like having a comment here, but I think it could be refined. I had some confusion reading `select` and interpreted it as a user action selecting something, but it's actually pre-selecting before the prompt. 

What do you think about something like this, Copilot assisted, revision:

```Go
// Populate the list of selectable assignees and their default selections.
// This logic maps the default assignees from `state` to the corresponding actors or users
// so that the correct display names are preselected in the prompt.
```

BagToad commented on pkg/cmd/pr/shared/survey.go:
```diff
@@ -219,6 +240,7 @@ func MetadataSurvey(p Prompt, io *iostreams.IOStreams, baseRepo ghrepo.Interface
 		milestones = append(milestones, m.Title)
 	}
 
+	// Prompt user for additional metadata based on selected fields
```
Thank you - I find these types of comments very helpful to leave behind in code like this that has rough lines between when one process starts and another ends. It's a small thing, but I think it adds up to less cognitive load reading the code.

BagToad commented on pkg/cmd/pr/shared/survey.go:
```diff
@@ -199,9 +202,27 @@ func MetadataSurvey(p Prompt, io *iostreams.IOStreams, baseRepo ghrepo.Interface
+	// and translate them to the appropriate actor / user.
 	var assignees []string
-	for _, u := range metadataResult.AssignableUsers {
-		assignees = append(assignees, u.DisplayName())
+	var assigneesDefault []string
+	if state.ActorAssignees {
+		for _, u := range metadataResult.AssignableActors {
+			assignees = append(assignees, u.DisplayName())
+
+			if slices.Contains(state.Assignees, u.Login()) {
+				assigneesDefault = append(assigneesDefault, u.DisplayName())
+			}
+		}
+	} else {
+		for _, u := range metadataResult.AssignableUsers {
+			assignees = append(assignees, u.DisplayName())
+
+			if slices.Contains(state.Assignees, u.Login()) {
+				assigneesDefault = append(assigneesDefault, u.DisplayName())
+			}
+		}
```
thought: This feels like there is an opportunity to collapse these into a single loop since the logic in each loop is the same aside from the set we're iterating through. I'll checkout the code and try it out - if it's fruitful I'll post a suggestion.

andyfeller commented on pkg/cmd/pr/shared/survey.go:
```diff
@@ -3,6 +3,7 @@ package shared
 import (
 	"errors"
 	"fmt"
```
Ignoring this suggestion as the `go.mod` is the authority on the version of Go required for the project.

andyfeller committed:
Improve comments from PR feedback

andyfeller committed:
Implement actor and user assignee tests

andyfeller commented on pkg/cmd/issue/create/create_test.go:
```diff
@@ -404,22 +413,23 @@ func Test_createRun(t *testing.T) {
-				"hasIssuesEnabled": true
-			} } }`))
+						{ "data": { "repository": {
+							"id": "REPOID",
+							"hasIssuesEnabled": true
+						} } }`))
 				r.Register(
 					httpmock.GraphQL(`mutation IssueCreate\b`),
 					httpmock.GraphQLMutation(`
-		{ "data": { "createIssue": { "issue": {
-			"URL": "https://github.com/OWNER/REPO/issues/12"
-		} } } }
-	`, func(inputs map[string]interface{}) {
+						{ "data": { "createIssue": { "issue": {
+							"URL": "https://github.com/OWNER/REPO/issues/12"
+						} } } }
+					`, func(inputs map[string]interface{}) {
 						assert.Equal(t, "title", inputs["title"])
 						assert.Equal(t, "body", inputs["body"])
 					}))
 			},
```
**note:** I have a hard time understanding test cases formatted this way, so I realigned the GraphQL JSON responses.  I would like to clean up some of the other cases if others find it helpful.

BagToad replied:
Always a fan of better indentation/readability over here 👍

andyfeller committed:
Refactor CopilotReplacer logic based on PR feedback

Thanks to @bagtoad, this commit refactors the argument to `NewCopilotReplacer(bool)` from being where this is used to what it effect is has.

Because there is already precedence for display name being `<login> (<name>)`, I worried there would be confusion for `Copilot (AI)` being display name for assignees and reviewers but `Copilot` when going to GitHub.com UI.  Instead, I renamed the argument based on whether the login is returned / replaced.

andyfeller committed:
Use constant for Copilot name

Based on PR feedback from @bagtoad, this commit creates a new constant for the Copilot bot name, which is used in the assignee / reviewer selection as well as replacing `@copilot` when going to GitHub.com UI

andyfeller committed:
Fix copilot replace test names

In refactoring how this parameter works, I forgot to update the related tests to match.

andyfeller committed:
Fix failing tests for Copilot replacement

BagToad reviewed (approved):
LGTM! 🚀 

Manual tests also look good 👍 

It's very cool you were able to swap to the shared `api.RepoMetadata`. I compared behaviors and I don't see anything jumping out as concerning byproducts of this change.

I responded to you in https://github.com/cli/cli/pull/11279/files#r2211313179 regarding whether to conditionalize the copilot replacer based on actor support, but I don't think a decision either way on that should block this.

danwulff commented:
FYI for those stumbling across this. As of Jul 29, this is not yet released to the Ubuntu runner (GitHub hosted runner): https://github.com/actions/runner-images/issues/12648
